File: scripts/inventory/LootTable.cs
using System;
using System.Collections.Generic;

namespace ProiectDeLegume.Scripts.Inventory;

public record LootEntry(string ItemId, int Weight);

public static class LootTables
{
    private static readonly Random Rng = new();

    private static readonly Dictionary<string, List<LootEntry>> Tables = new()
    {
        ["trash_can"] = new()
        {
            new("newspaper", 30), new("empty_can", 25), new("old_shirt", 10),
            new("canned_beans", 8), new("water_bottle", 5), new("screwdriver", 5),
            new("flashlight", 2)
        },
        ["kitchen"] = new()
        {
            new("canned_beans", 25), new("kitchen_knife", 15), new("matches", 15),
            new("water_bottle", 10), new("empty_can", 10)
        },
        ["wardrobe"] = new()
        {
            new("old_shirt", 40), new("flashlight", 5), new("matches", 5)
        },
        ["medkit"] = new()
        {
            new("bandage", 60), new("water_bottle", 10)
        },
        ["garage"] = new()
        {
            new("screwdriver", 30), new("flashlight", 15), new("matches", 10)
        }
    };

    public static List<ItemStack> Generate(string containerType)
    {
        if (!Tables.TryGetValue(containerType, out var table)) return new();

        int totalWeight = 0;
        foreach (var entry in table) totalWeight += entry.Weight;

        int itemCount = Rng.Next(1, 5);
        var result = new List<ItemStack>();

        for (int i = 0; i < itemCount; i++)
        {
            int roll = Rng.Next(totalWeight);
            int cumulative = 0;
            foreach (var entry in table)
            {
                cumulative += entry.Weight;
                if (roll < cumulative)
                {
                    var existing = result.Find(s => s.Def.Id == entry.ItemId && s.CanAdd());
                    if (existing != null)
                        existing.Count++;
                    else
                        result.Add(new ItemStack(ItemRegistry.Get(entry.ItemId)));
                    break;
                }
            }
        }

        return result;
    }
}

File: src/commonMain/kotlin/inventory/LootTable.kt
package inventory

import kotlin.random.Random

/**
 * Loot table entry: item id + weight (probability).
 */
data class LootEntry(val itemId: String, val weight: Int)

/**
 * Loot tables for different container types.
 */
object LootTables {
    private val tables = mapOf(
        "trash_can" to listOf(
            LootEntry("newspaper", 30),
            LootEntry("empty_can", 25),
            LootEntry("old_shirt", 10),
            LootEntry("canned_beans", 8),
            LootEntry("water_bottle", 5),
            LootEntry("screwdriver", 5),
            LootEntry("flashlight", 2)
        ),
        "kitchen" to listOf(
            LootEntry("canned_beans", 25),
            LootEntry("kitchen_knife", 15),
            LootEntry("matches", 15),
            LootEntry("water_bottle", 10),
            LootEntry("empty_can", 10)
        ),
        "wardrobe" to listOf(
            LootEntry("old_shirt", 40),
            LootEntry("flashlight", 5),
            LootEntry("matches", 5)
        ),
        "medkit" to listOf(
            LootEntry("bandage", 60),
            LootEntry("water_bottle", 10)
        ),
        "garage" to listOf(
            LootEntry("screwdriver", 30),
            LootEntry("flashlight", 15),
            LootEntry("matches", 10)
        )
    )

    /**
     * Generate loot for a container type. Returns 1-4 random items.
     */
    fun generate(containerType: String): List<ItemStack> {
        val table = tables[containerType] ?: return emptyList()
        val totalWeight = table.sumOf { it.weight }
        val itemCount = Random.nextInt(1, 5) // 1 to 4 items
        val result = mutableListOf<ItemStack>()

        repeat(itemCount) {
            val roll = Random.nextInt(totalWeight)
            var cumulative = 0
            for (entry in table) {
                cumulative += entry.weight
                if (roll < cumulative) {
                    // Try to stack with existing
                    val existing = result.find { it.def.id == entry.itemId && it.canAdd() }
                    if (existing != null) {
                        existing.count++
                    } else {
                        result.add(ItemStack(ItemRegistry.get(entry.itemId)))
                    }
                    break
                }
            }
        }

        return result
    }
}
